tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp

```diff
@@ -64,15 +64,74 @@ void EditorDiagConsumer::getAllDiagnostics(
 
   Result.append(InvalidLocDiagnostics.begin(), InvalidLocDiagnostics.end());
 
-  // Note: we cannot reuse InputBufIds because there may be diagnostics outside
-  // the inputs.  Instead, sort the extant buffers.
+  // Note: We cannot use the input buffer IDs because there may be diagnostics
+  // outside the inputs. Instead, sort the extant buffers.
   auto bufferIDs = getSortedBufferIDs(BufferDiagnostics);
   for (unsigned bufferID : bufferIDs) {
     const auto &diags = BufferDiagnostics[bufferID];
     Result.append(diags.begin(), diags.end());
   }
 }
 
+/// Retrieve the raw range from a range, if it exists in the provided buffer.
+/// Otherwise returns \c None.
+static Optional<RawCharSourceRange> getRawRangeInBuffer(CharSourceRange range,
+                                                        unsigned bufferID,
+                                                        SourceManager &SM) {
+  if (!range.isValid() || !SM.findBufferContainingLoc(range.getStart()))
+    return None;
+
+  unsigned offset = SM.getLocOffsetInBuffer(range.getStart(), bufferID);
+  unsigned length = range.getByteLength();
+  return {{offset, length}};
+}
+
+BufferInfoSharedPtr EditorDiagConsumer::getBufferInfo(
+    StringRef FileName, Optional<unsigned> BufferID, swift::SourceManager &SM) {
+  // NOTE: Using StringRef as a key here relies on SourceMgr using const char*
+  // as buffer identifiers. This is fast, but may be brittle.  We can always
+  // switch over to using a StringMap. Note that the logic in
+  // SerializedDiagnosticConsumer::getEmitFile will also need changing.
+  auto Result = BufferInfos.find(FileName);
+  if (Result != BufferInfos.end())
+    return Result->second;
+
+  Optional<std::string> GeneratedFileText;
+  Optional<BufferInfo::OriginalLocation> OriginalLocInfo;
+
+  // If we have a generated buffer, we need to include the source text, as
+  // clients otherwise won't be able to access to it.
+  if (BufferID) {
+    if (auto Info = SM.getGeneratedSourceInfo(*BufferID)) {
+      // Don't include this information for replaced function body buffers, as
+      // they'll just be referencing the original file.
+      // FIXME: Does this mean that the location information we'll be giving
```

Comment:
`ReplacedFunctionBody` only happens for completion/cursor info at the moment. So we should never end up with that case here. You could add an assertion if you want.

Reply:
We actually hit this case when doing multiple `compile` requests

Reply:
Oh interesting. @rintaro despite you just talking `compile` I didn't think about it here. It does seem like we would be messing up diagnostics then, I'm unclear whether we actually care though - I don't believe we send them back.

Reply:
Yes. But I didn't think much about diagnostics for `compile` requests. In our intended use cases, I thought we only wanted to know if there's any error or not (succeeded or not). But anyway `compile` request hasn't been maintained well, and probably requires major refinements.